routing/router.go

```diff
@@ -1668,119 +1743,209 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 
 	timeoutChan := time.After(payAttemptTimeout)
 
+	paymentHash := payment.PaymentHash
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
-	var lastError error
-	for {
+	var (
+		lastError error
+	)
+
+	// sendNewAttempt is a helper method that creates and sends a new
+	// payment to the switch. We use this when no payment has been sent to
+	// the switch already, or the previous attempt failed.
+	sendNewAttempt := func() error {
 		// Before we attempt this next payment, we'll check to see if
 		// either we've gone past the payment attempt timeout, or the
 		// router is exiting. In either case, we'll stop this payment
 		// attempt short.
 		select {
 		case <-timeoutChan:
+			// Mark the payment as failed.
+			err := a.Fail()
+			if err != nil {
+				return err
+			}
+
 			errStr := fmt.Sprintf("payment attempt not completed "+
 				"before timeout of %v", payAttemptTimeout)
 
-			return [32]byte{}, nil, newErr(
-				ErrPaymentAttemptTimeout, errStr,
-			)
+			// Terminal state, return.
+			return newErr(ErrPaymentAttemptTimeout, errStr)
 
 		case <-r.quit:
-			return [32]byte{}, nil, fmt.Errorf("router shutting down")
+			// The payment will be resumed from the current state
+			// after restart.
+			return ErrRouterShuttingDown
 
 		default:
 			// Fall through if we haven't hit our time limit, or
 			// are expiring.
 		}
 
+		// Create a new payment attempt from the given payment session.
 		route, err := paySession.RequestRoute(
 			payment, uint32(currentHeight), finalCLTVDelta,
 		)
 		if err != nil {
 			// If we're unable to successfully make a payment using
-			// any of the routes we've found, then return an error.
+			// any of the routes we've found, then mark the payment
+			// as permanently failed.
+			saveErr := a.Fail()
+			if saveErr != nil {
+				return saveErr
+			}
+
+			// If there was an error already recorded for this
+			// payment, we'll return that.
 			if lastError != nil {
```

Comment:
I think we only want to return `lastError` in case `err == ErrNoPathFound`. Other (unexpected) path finding errors should have a high precedence over the `lastError` I think.

Reply:
Can be done in later PR, attempting to keep the RPC behavior as unchanged as possible in this PR.